src/lib.rs

```diff
@@ -1410,6 +1480,356 @@ where
+                }
+                Err(DeviceCodeAction::IncreaseIntervalThenRetry) => {
+                    // Increase the interval, then wait that long.
+                    interval = interval + Duration::from_secs(5);
+                }
+                Err(DeviceCodeAction::DoubleIntervalThenRetry) => {
+                    // Double the interval, then wait that long.
+                    interval = interval.checked_mul(2).ok_or(RequestTokenError::Other(
+                        "Failed to increase interval".to_string(),
+                    ))?;
+                }
+                Err(DeviceCodeAction::NoFurtherRequests(req)) => {
+                    break token_response(req)
+                }
+            };
+
+            // Actually sleep here.
+            std::thread::sleep(interval);
+            elapsed = elapsed + interval;
+        }
+    }
+
+    ///
+    /// Asynchronously sends the request to the authorization server and awaits a response.
+    ///
+    pub async fn request_async<C, F, RE>(
+        self,
+        http_client: C,
+    ) -> Result<TR, RequestTokenError<RE, TE>>
+    where
+        C: FnOnce(HttpRequest) -> F + Copy,
+        F: Future<Output = Result<HttpResponse, RE>>,
+        RE: Error + 'static,
+    {
+        let details = self.device_authorization_details.ok_or_else(|| {
+            RequestTokenError::Other("no device authorization details provided".to_string())
+        })?;
+
+        let mut elapsed = Duration::new(0, 0);
+        let mut interval = details.interval().clone();
+
+        loop {
+            if elapsed > details.expires_in() {
+                return Err(RequestTokenError::Other("Device code expired".to_string()));
+            }
+
+            // Prepare the request.
+            match http_client(self.prepare_request()?)
+                .await
+                .map_err(|_| DeviceCodeAction::DoubleIntervalThenRetry)
+                .and_then(device_token_action)
+            {
+                Ok(http_response) => break token_response(http_response),
+                Err(DeviceCodeAction::Retry) => {
+                    // Wait for the current interval.
+                }
+                Err(DeviceCodeAction::IncreaseIntervalThenRetry) => {
+                    // Increase the interval, then wait that long.
+                    interval = interval + Duration::from_secs(5);
+                }
+                Err(DeviceCodeAction::DoubleIntervalThenRetry) => {
+                    // Double the interval, then wait that long.
+                    interval = interval.checked_mul(2).ok_or(RequestTokenError::Other(
+                        "Failed to increase interval".to_string(),
+                    ))?;
+                }
+                Err(DeviceCodeAction::NoFurtherRequests(req)) => {
+                    break token_response(req)
+                }
+            };
+
+            // Use async-std to sleep asynchronously.
+            async_std::task::sleep(interval).await;
```

Comment:
We definitively have to think about this one. I'm not sure how well this works when using tokio (or any other future async runtime provider). Worst case means having to deal with extra flags and being incompatible to new libraries by default.

Reply:
Yeah, I thought this might be the most controversial bit. Happy to hear alternative solutions (use a blocking sleep, don't expose an async method yet)

Reply:
great point! in addition to the concerns @mettke raised, having this code sleep means that both the crate itself, as well as downstream clients, can't easily write regression tests that don't block. any time we're dealing with an interface that involves time (either sleeping or returning the current time), it's useful to structure the interface in a way that allows time to be mocked easily.

to that end, I think we should follow @mettke's suggestion and have the caller provide the sleeper function, which could use `async_std`, some other implementation, or be a no-op mock function for testing.

for `request_async`, let's add a `sleep_fn` parameter with the signature `Fn(Duration) -> Future<Output=()>` or something to that effect. most users can just pass `async_std::task::sleep` directly as that argument, which we can show in the documentation. this allows us to remove the `async_std` dependency from this crate (though we probably need it as a dev dependency for tests and example code).

for the sync version `request`, I think the `sleep_fn` signature would be `Fn(Duration) -> ()`. for that case, it's just mockability that we care about.

Reply:
Seems reasonable! Will fix up.

Reply:
Seems reasonable! Will fix up.